necessary tests for progbufsize | Skip CSR read if insufficient progbuf in RISC-V targets

Biancaa-R opened the pull request:
This PR resolves this issue https://github.com/riscv-collab/riscv-openocd/issues/1264
These functions currently do not exit early if program the program buffer in the target is insufficient:

    csr_read_progbuf()
    fpr_read_progbuf()
    csr_write_progbuf()
    fpr_write_progbuf()
 This pr resolves this issue using the pre existing function:
`if (!has_sufficient_progbuf(target, 2))`

Add a check in CSR read functions to skip the operation when the target's
progbuf is smaller than required. This prevents errors on targets with
limited progbuf size and ensures safer debugging operations.

- Added `has_sufficient_progbuf` checks to CSR read functions.
- Log debug message when skipping due to insufficient progbuf.
- Improves compatibility with smaller RISC-V targets.

<img width="677" height="453" alt="image" src="https://github.com/user-attachments/assets/7b704148-494e-4c79-b4b6-c2f894e110ce" />
* checks for the min size

Successful configuring:
<img width="1314" height="736" alt="image" src="https://github.com/user-attachments/assets/f312a1e9-fa8b-492c-9fc8-3b7537d11fe6" />
Successful build ,after build the version was also tested to check working:
<img width="1314" height="736" alt="image" src="https://github.com/user-attachments/assets/b8aa0209-c9bd-4fd9-98ed-b1f1be9f6feb" />
with `./src/openocd --version`

Commit:
added the necessary tests for progbufsize

Added additional debug logs to verify correct retrieval and printing
of progbufsize in various code paths within riscv-013.c. This helps
in debugging hardware configurations that expose custom program buffer
sizes.

Signed-off-by: Biancaa Ramesh <biancaa2210329@ssn.edu.in>

Biancaa-R commented:
This code built and ran without issues ,In case of any needed modification of functionality ,please let me know ,Ill do it :)

en-sc commented on src/target/riscv/riscv-013.c:
```diff
@@ -1385,6 +1385,11 @@ static int fpr_read_progbuf(struct target *target, uint64_t *value,
 	assert(target->state == TARGET_HALTED);
 	assert(number >= GDB_REGNO_FPR0 && number <= GDB_REGNO_FPR31);
+	if (!has_sufficient_progbuf(target, 2)) {
+		LOG_DEBUG("Skipping FPR read: insufficient progbuf (size=%d)",
```
[here and around] Please, use `LOG_TARGET_DEBUG(...)` to give a bit more context:
```suggestion
		LOG_TARGET_DEBUG(target, "Skipping FPR read: insufficient progbuf (size=%d)",
```
Also `progbufsize` is unsigned, so the correct format specifier is `%u` (though this is undoubtedly a nitpick).

Biancaa-R replied:
Changed it in the updated version.

en-sc reviewed (changes requested):
@Biancaa-R, thank you for submitting the patch!

Please, take a look at Checkpatch failures.
The first `ERROR:COMMIT_MESSAGE:` can be quite cryptic -- the commit message is split into the title and the description with an empty line in-between.
Please see https://git-scm.com/docs/git-commit#_discussion for more.

The other errors seem descriptive enough.

You can run the script locally with:
```
> tools/scripts/checkpatch.pl -g HEAD
```

Also, what about `vtype_write_progbuf()`/`vl_write_progbuf()`?

Commit:
riscv: Add have_progbuf_size check in vtype,vl writes and crs,fpr read-write

Added a check for have_progbuf_size before attempting to execute. This prevents program buffer access on targets
that do not support it, improving robustness and avoiding potential
execution failures when the program buffer is not available or too small.

Signed-off-by: Biancaa Ramesh <biancaa2210329@ssn.edu.in>

en-sc commented on src/target/riscv/riscv-013.c:
```diff
@@ -1536,6 +1554,11 @@ static int vtype_write_progbuf(struct target *target, riscv_reg_t value)
 {
 	assert(target->state == TARGET_HALTED);
+	/* Ensure program buffer is large enough for 2 instructions */
+	if (!has_sufficient_progbuf(target, 2)) {
```
```suggestion
	if (!has_sufficient_progbuf(target, 3)) {
```

The sequence is:
```
csrr ...
vsetvl ...
ebreak
```

en-sc commented on src/target/riscv/riscv-013.c:
```diff
@@ -1515,6 +1528,11 @@ static int fpr_write_progbuf(struct target *target, enum gdb_regno number,
 {
 	assert(target->state == TARGET_HALTED);
+	/* Ensure program buffer is large enough for 2 instructions */
+	if (!has_sufficient_progbuf(target, 2)) {
```
```suggestion
	if (!has_sufficient_progbuf(target, 3)) {
```
(Same as for `vl`.)

Biancaa-R replied:
Oh as For vector register (v1) or vector type (vtype) operations we need to see 3 instructions ,right !
Ill fix it .Thanks @en-sc

Commit:
riscv: Modified the necessary tests for progbuf size

check the progbuf sizes to 2,3 and skip if there is in sufficient progbuf

Signed-off-by: Biancaa Ramesh <biancaa2210329@ssn.edu.in>

Biancaa-R commented:
Sorry about the indentation for comments, I used 
<img width="1215" height="47" alt="image" src="https://github.com/user-attachments/assets/6e543356-0b41-4947-9a73-204803180d95" />
To fix indentation issues. I hope now its in the way you have expected .

en-sc commented on src/target/riscv/riscv-013.c:
```diff
@@ -1,9 +1,9 @@
- * Support for RISC-V, debug version 0.13, which is currently (2/4/17) the
- * latest draft.
- */
+	* Support for RISC-V, debug version 0.13, which is currently (2/4/17) the
```
This is not needed, it's okay to have a space here.

en-sc replied:
Now the space is missing)

Biancaa-R replied:
> Now the space is missing)

Ok Ill add it back 🦅
